qiskit_experiments/analysis/curve_fitting.py

```diff
@@ -0,0 +1,291 @@
+        weights: Optional, a 1D float list of weights :math:`w_k` for each
+                 component function :math:`f_k`.
+        bounds: Optional, lower and upper bounds for optimization
+                parameters.
+        kwargs: additional kwargs for scipy.optimize.curve_fit.
+
+    Returns:
+        AnalysisResult: result containing `popt` the optimal fit parameters,
+                        `popt_err` the standard error estimates popt,
+                        `pcov` the covariance matrix for the fit,
+                        `chisq` the chi-squared parameter of fit.
+                        `xrange` the range of xdata values used for fit.
+
+    Raises:
+        QiskitError: if the number of y-values in the dataset is not
+                     greater than the number of parameters in the
+                     fit function.
+    """
+    num_funcs = len(funcs)
+
+    # Get positions for indexes data sets
+    series = np.asarray(series, dtype=int)
+    idxs = [series == i for i in range(num_funcs)]
+
+    # Combine weights and sigma for transformation
+    if weights is None:
+        wsigma = sigma
+    else:
+        wsigma = np.zeros(ydata.size)
+        if sigma is None:
+            for i in range(num_funcs):
+                wsigma[idxs[i]] = 1 / np.sqrt(weights[i])
+        else:
+            for i in range(num_funcs):
+                wsigma[idxs[i]] = sigma[idxs[i]] / np.sqrt(weights[i])
+
+    # Define multi-objective function
+    def f(x, *params):
+        y = np.zeros(x.size)
+        for i in range(num_funcs):
+            xi = x[idxs[i]]
+            yi = funcs[i](xi, *params)
+            y[idxs[i]] = yi
+        return y
+
+    # Run linearized curve_fit
+    analysis_result = curve_fit(f, xdata, ydata, p0, sigma=wsigma, bounds=bounds, **kwargs)
+
+    return analysis_result
+
+
+def process_curve_data(
+    data: List[Dict[str, any]], data_processor: Callable, x_key: str = "xval", **filters
+) -> Tuple[np.ndarray, np.ndarray, np.ndarray]:
+    """Return tuple of arrays (x, y, sigma) data for curve fitting.
+
+    Args
+        data: list of circuit data dictionaries containing counts.
+        data_processor: callable for processing data to y, sigma
+        x_key: key for extracting xdata value from metadata (Default: "xval").
+        filters: additional kwargs to filter metadata on.
+
+    Returns:
+        tuple: ``(x, y, sigma)`` tuple of arrays of x-values,
+               y-values, and standard deviations of y-values.
+
+    Raises:
+        QiskitError: if input data is not level-2 measurement.
+    """
+    filtered_data = filter_data(data, **filters)
+    size = len(filtered_data)
+    xdata = np.zeros(size, dtype=float)
+    ydata = np.zeros(size, dtype=float)
+    ydata_var = np.zeros(size, dtype=float)
+
+    for i, datum in enumerate(filter_data):
+        metadata = datum["metadata"]
+        xdata[i] = metadata[x_key]
+        y_mean, y_var = data_processor(datum)
```

Comment:
The data process should be aware of whether `sigma` is absolute

Reply:
This one is not a mistake. The computation of `sigma` is different for different values of `absolute_sigma`.

Reply:
@eggerdj Note the previous comment because it's relevant to #22

Reply:
The choice of setting absolute_sigma = False/True should depend on how sigma was calculated, not the other way around. Absolute sigma is a property of curve fit function for how it calculates `pcov` (and hence how we deduce `popt_err`). According to scipy docs this is `pcov(absolute_sigma=False) = pcov(absolute_sigma=True) * chisq(popt)/(M-N)`

By defaulting to True we are assuming that sigma is usually specified in the same units of y, so in that sense we should make sure any data processors calculate sigma in the same units as y by default as well.